include/deal.II/lac/solver_idr.h

```diff
@@ -0,0 +1,495 @@
+ */
+template <class VectorType = Vector<double>>
+class SolverIDR : public SolverBase<VectorType>
+{
+public:
+  /**
+   * Structure for storing additional data needed by the solver.
+   */
+  struct AdditionalData
+  {
+    explicit AdditionalData(const unsigned int s = 2)
+      : s(s)
+    {}
+
+    const unsigned int s;
+  };
+
+  /**
+   * Constructor.
+   */
+  SolverIDR(SolverControl &           cn,
+            VectorMemory<VectorType> &mem,
+            const AdditionalData &    data = AdditionalData(2));
+
+  /**
+   * Constructor. Use an object of type GrowingVectorMemory as a default to
+   * allocate memory.
+   */
+  SolverIDR(SolverControl &cn, const AdditionalData &data = AdditionalData(2));
+
+  /**
+   * Virtual destructor.
+   */
+  virtual ~SolverIDR() override = default;
+
+  /**
+   * Solve the linear system <code>Ax=b</code> for x.
+   */
+  template <typename MatrixType, typename PreconditionerType>
+  void
+  solve(const MatrixType &        A,
+        VectorType &              x,
+        const VectorType &        b,
+        const PreconditionerType &preconditioner);
+
+protected:
+  /**
+   * Interface for derived class. This function gets the current iteration
+   * vector, the residual and the update vector in each step. It can be used
+   * for graphical output of the convergence history.
+   */
+  virtual void
+  print_vectors(const unsigned int step,
+                const VectorType & x,
+                const VectorType & r,
+                const VectorType & d) const;
+
+private:
+  /**
+   * Additional solver parameters.
+   */
+  AdditionalData additional_data;
+};
+
+/*@}*/
+/*------------------------- Implementation ----------------------------*/
+
+#ifndef DOXYGEN
+
+
+namespace internal
+{
+  namespace SolverIDRImplementation
+  {
+    template <class VectorType>
+    inline TmpVectors<VectorType>::TmpVectors(const unsigned int        s_param,
+                                              VectorMemory<VectorType> &vmem)
+      : mem(vmem)
+      , data(s_param)
+    {}
+
+
+
+    template <class VectorType>
+    inline VectorType &TmpVectors<VectorType>::
+                       operator[](const unsigned int i) const
+    {
+      Assert(i < data.size(), ExcIndexRange(i, 0, data.size()));
+
+      Assert(data[i] != nullptr, ExcNotInitialized());
+      return *data[i];
+    }
+
+
+
+    template <class VectorType>
+    inline VectorType &
+    TmpVectors<VectorType>::operator()(const unsigned int i,
+                                       const VectorType & temp)
+    {
+      Assert(i < data.size(), ExcIndexRange(i, 0, data.size()));
+      if (data[i] == nullptr)
+        {
+          data[i] = std::move(typename VectorMemory<VectorType>::Pointer(mem));
+          data[i]->reinit(temp);
+        }
+      return *data[i];
+    }
+
+
+
+    template <class VectorType>
+    unsigned int
+    TmpVectors<VectorType>::size() const
+    {
+      return (data.size() > 0 ? data.size() - 1 : 0);
```

Comment:
I don't understand this logic. Can you explain?

Reply:
This is actually left over from the GMRES implementation. I do not need this function and will delete it.